### PR TITLE
Feature/set id root claim

### LINKED ISSUE #17: New claim definition: SetRootKey

### LINKED ISSUE #18: New claim definition: Authorize KSign

### LINKED ISSUE #19: New claim definition: Basic

### LINKED ISSUE #20: New claim definition: AssignName

### PR DESCRIPTION
**Pull request by krlosMata**
New claims definition:
  - assign-name
  - authorize-ksign
  - basic
  - set-root-key

Structure `Elements` defined to represent claim object

close #17, close #18, close #19, close #20

### CODE REVIEW

File: src/claim/assign-name/assign-name.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
+const Claim = require('../claim');
+const utils = require('../../utils');
+
+/**
+ * @param  {Uint32} Version
```
Comment (cjgarde) [inline src/claim/assign-name/assign-name.js]: Add a short description of the class please, not only parameters... Those parameters should go in the constructor comments

File: src/claim/assign-name/assign-name.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+ * @param  {String} Hash name
+ * @param  {String} Address identifier
+ */
+class AssignName {
+  constructor(_version = 0, _hashName = '', _id = '') {
```
Comment (cjgarde) [inline src/claim/assign-name/assign-name.js]: Why lodash in the arguments? Are not private vars of the class, are arguments.

File: src/claim/assign-name/assign-name.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    this.structure = {
+      claimType: utils.hashBytes('iden3.claim.assign_name').slice(24, 32),
+      version: versionBuff,
+      hashName: utils.hashBytes(_hashName).fill(0, 0, 1),
+      id: utils.hexToBytes(_id),
+    };
+  }
+
+  /**
+   * @returns {Object Elements} Element representation of the claim
```
Comment (cjgarde) [inline src/claim/assign-name/assign-name.js]: Add a short description of the function please

File: src/claim/assign-name/assign-name.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+   */
+  elements() {
+    const element = new Claim.Elements();
+    let endIndex = element.e3.length;
```
Comment (cjgarde) [inline src/claim/assign-name/assign-name.js]: What does e3 mean? Could you provide some comments about what is the purpose of these lines of code?

File: src/claim/assign-name/assign-name.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    let startIndex = endIndex - this.structure.claimType.length;
+    // e3 element composition
+    element.e3.fill(this.structure.claimType, startIndex, endIndex);
```
Comment (cjgarde) [inline src/claim/assign-name/assign-name.js]: could

File: src/claim/assign-name/assign-name.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    endIndex = element.e1.length;
+    startIndex = element.e1.length - this.structure.id.length;
+    element.e1.fill(this.structure.id, startIndex, endIndex);
+    // e0 remains as empty value
+    return element;
+  }
+}
+
+/**
+ * @param  {Object Elements} Representation of the claim
```
Comment (cjgarde) [inline src/claim/assign-name/assign-name.js]: Add a short description of the function please

File: src/claim/assign-name/assign-name.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+ * @returns {Claim class} Claim object
```
Comment (cjgarde) [inline src/claim/assign-name/assign-name.js]: I think that is @returns {Object} Claim

File: src/claim/assign-name/assign-name.js (hunk 6 above)
Comment (cjgarde) [inline src/claim/assign-name/assign-name.js]: @param {Object} elements - representation of the claims

File: src/claim/assign-name/assign-name.js [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+ */
+const parseFromElements = function (elements) {
```
Comment (cjgarde) [inline src/claim/assign-name/assign-name.js]: Please, name the function for a better debug:
 const parseFromElements = function parseElements(elements) {...}

File: src/claim/authorize-ksign/authorize-ksign.js [hunk 9]
```diff
@@ -0,0 +1,72 @@
+const Claim = require('../claim');
+const utils = require('../../utils');
+
+/**
```
Comment (cjgarde) [inline src/claim/authorize-ksign/authorize-ksign.js]: Please, short description of the class and the parameters go to the constructor 
(http://usejsdoc.org/howto-es2015-classes.html)

File: src/claim/authorize-ksign/authorize-ksign.js [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+ * @param  {Uint32} Version
+ * @param  {Bool} Signe
+ * @param  {String} Ax
+ * @param  {String} Ay
+ */
+class AuthorizeKSign {
+  constructor(_version = 0, _signe = false, _ax = '', _ay = '') {
```
Comment (cjgarde) [inline src/claim/authorize-ksign/authorize-ksign.js]: Why lodash in the arguments?

File: src/claim/authorize-ksign/authorize-ksign.js [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    const versionBuff = Buffer.alloc(4);
```
Comment (cjgarde) [inline src/claim/authorize-ksign/authorize-ksign.js]: We should have in the comments why are you allocating 4 bytes, 1 bytes, etc... 
Or the claim structure commented ....

File: src/claim/authorize-ksign/authorize-ksign.js [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+    const signeBuff = Buffer.alloc(1);
+    signeBuff.writeUInt8(_signe);
```
Comment (cjgarde) [inline src/claim/authorize-ksign/authorize-ksign.js]: signe? or signed?

File: src/claim/authorize-ksign/authorize-ksign.js [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    versionBuff.writeUInt32BE(_version);
+    this.structure = {
+      claimType: utils.hashBytes('iden3.claim.authorize_k_sign').slice(24, 32),
+      version: versionBuff,
+      signe: signeBuff,
```
Comment (cjgarde) [inline src/claim/authorize-ksign/authorize-ksign.js]: signe or signed?

File: src/claim/authorize-ksign/authorize-ksign.js [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+      ax: utils.hexToBytes(_ax),
+      ay: utils.hexToBytes(_ay),
+    };
+  }
+
+  /**
```
Comment (cjgarde) [inline src/claim/authorize-ksign/authorize-ksign.js]: Short description of the method, please

File: src/claim/authorize-ksign/authorize-ksign.js [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+   * @returns {Object Elements} Element representation of the claim
```
Comment (cjgarde) [inline src/claim/authorize-ksign/authorize-ksign.js]: @returns {Object} element

File: src/claim/authorize-ksign/authorize-ksign.js [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+   */
+  elements() {
+    const element = new Claim.Elements();
+    let endIndex = element.e3.length;
+    let startIndex = endIndex - this.structure.claimType.length;
+    // e3 element composition
```
Comment (cjgarde) [inline src/claim/authorize-ksign/authorize-ksign.js]: Composition can be don in a loop, right? :)

File: src/claim/authorize-ksign/authorize-ksign.js [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+    endIndex = element.e2.length;
+    startIndex = element.e2.length - this.structure.ay.length;
+    element.e2.fill(this.structure.ay, startIndex, endIndex);
+    // e1 remains as empty value
+    // e0 remains as empty value
+    return element;
+  }
+}
+
+/**
```
Comment (cjgarde) [inline src/claim/authorize-ksign/authorize-ksign.js]: short description of the function, please

File: src/claim/authorize-ksign/authorize-ksign.js [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+ * @param  {Object Elements} Representation of the claim
```
Comment (cjgarde) [inline src/claim/authorize-ksign/authorize-ksign.js]: @param {Object) element ... or can define your type like http://usejsdoc.org/tags-typedef.html

File: src/claim/authorize-ksign/authorize-ksign.js [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+ * @returns {Claim class} Claim object
```
Comment (cjgarde) [inline src/claim/authorize-ksign/authorize-ksign.js]: @returns {Object} Claim or define your type like http://usejsdoc.org/tags-typedef.html

File: src/claim/authorize-ksign/authorize-ksign.js [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+ */
+const parseFromElements = function (elements) {
```
Comment (cjgarde) [inline src/claim/authorize-ksign/authorize-ksign.js]: Please, name the function for better debugging, if not an "Anonymous function" will appear 
const parseFromElements = function parseElements(elements) ...

File: src/claim/authorize-ksign/authorize-ksign.js [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+  const claim = new AuthorizeKSign();
+  // Parse e3
+  claim.structure.claimType = elements.e3.slice(24, 32);
+  claim.structure.version = elements.e3.slice(20, 24);
+  claim.structure.signe = elements.e3.slice(19, 20);
+  claim.structure.ax = elements.e3.slice(3, 19);
```
Comment (cjgarde) [inline src/claim/authorize-ksign/authorize-ksign.js]: what does ax mean? ay? 
Could you comment the structure please?

**Comment by krlosMata** [comment]
Add several comments and information about the four new claims type

**Comment by krlosMata** [comment]
- Create claim `Factory` class to create claims
- Modify claim structure in order to create claim from a claim factory
- Update comments according code structure
- Create `Entry` class to add a layer under claim structure
- Modify claim test to fulfill above modifications

File: src/claim/claim.js [hunk 22]
```diff
@@ -66,11 +80,11 @@ class GenericClaim {
...
-      uint32ToEthBytes(this.claim.baseIndex.indexLength),
+      utils.uint32ToEthBytes(this.claim.baseIndex.indexLength),
     ]);
     b = Buffer.concat([
       b,
-      uint32ToEthBytes(this.claim.baseIndex.version),
+      utils.uint32ToEthBytes(this.claim.baseIndex.version),
     ]);
     b = Buffer.concat([b, this.claim.extraIndex.data]);
     b = Buffer.concat([b, this.claim.data]);
```
[the remark this answers is not in the document]
Comment (krlosMata) [reply src/claim/claim.js]: This issue is written new in new claim definition

File: src/claim/claim.js [hunk 23]
```diff
@@ -145,21 +159,21 @@ class AuthorizeKSignClaim {
     b = Buffer.concat([b, this.claim.baseIndex.type]);
```
[the remark this answers is not in the document]
Comment (krlosMata) [reply src/claim/claim.js]: This issue is written new in new claim definition

File: src/claim/claim.js [hunk 24]
```diff
@@ -183,17 +197,17 @@ class AuthorizeKSignClaim {
  */
 const parseAuthorizeKSignClaim = function (b) {
```
[the remark this answers is not in the document]
Comment (krlosMata) [reply src/claim/claim.js]: This function is outdated and it has to be removed after new claim integration

**Pull request merged**